### PR TITLE
MF-52  Create a Config Library skeleton

### PR DESCRIPTION
**Pull request by brandones**
This is the first draft of the config library that sort of does the minimum of what's expected -- takes schemas and configs and returns configs.

This also attempts to change over to typescript. Sorry for the sloppy use of git.

### CODE REVIEW

File: src/openmrs-esm-module-config.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  for (let key of Object.keys(schema)) {
+    if (! providedConfig.hasOwnProperty(key)) {
+      providedConfig[key] = schema[key]["default"];
+    }
+  }
+
+  return providedConfig;
+}
+
+export function clearAll() {
```
Comment (brandones) [inline src/openmrs-esm-module-config.ts]: This is just for testing. I'm sure there's a better way about this -- please suggest it if you know one.

File: package.json [hunk 2]
```diff
@@ -7,7 +7,8 @@
     "clean": "rimraf dist",
     "start": "webpack-dev-server --mode=development",
     "test": "jest",
```
Comment (brandones) [inline package.json]: Jest is complaining that it doesn't understand typescript. Something, probably involving babel, is not wired up correctly.

File: webpack.config.js [hunk 3]
```diff
@@ -17,17 +18,21 @@ module.exports = {
...
+        test: /\.m?(js|ts|tsx)$/,
+        exclude: /(node_modules|bower_components)/,
+        use: "babel-loader"
       }
     ]
   },
+  resolve: {
+    extensions: [".ts", ".js", ".tsx", ".jsx"]
+  },
+  externals: ["openmrs-config"],
```
Comment (joeldenning) [inline webpack.config.js]: What is `openmrs-config`?

File: src/openmrs-esm-module-config.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+// An object with module names for keys and schemas for values.
+const schemas = {};
+
+export function defineConfigSchema(moduleName, schema) {
+  console.log("defineConfigSchema received schema for " + moduleName + ": " + JSON.stringify(schema));
+  schemas[moduleName] = schema;
+}
+
+export function provide(config) {
+  console.log("provide recieved config " + JSON.stringify(config));
```
Comment (joeldenning) [inline src/openmrs-esm-module-config.ts]: i before e, except after c

😄

File: package.json [hunk 5]
```diff
@@ -19,7 +21,7 @@
   ],
   "husky": {
     "hooks": {
-      "pre-commit": "pretty-quick --staged && npm run lint && npm test"
+      "pre-commit": "pretty-quick --staged && npm run lint && npm test && npm run typescript"
```
Comment (joeldenning) [inline package.json]: If you want a perf boost to the pre-commit hook, see [here](https://github.com/openmrs/openmrs-esm-api/blob/954a83e1e19b145f4bb75a1c610de483150ef530/package.json#L21)

File: webpack.config.js [hunk 6]
```diff
@@ -1,14 +1,15 @@
 const path = require("path");
+const CleanWebpackPlugin = require("clean-webpack-plugin").CleanWebpackPlugin;
+const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
 
 module.exports = {
-  entry: path.resolve(__dirname, "src/openmrs-esm-module-config.js"),
+  entry: path.resolve(__dirname, "src/openmrs-esm-module-config.ts"),
```
[the remark this answers is not in the document]
Comment (brandones) [reply webpack.config.js]: Joel helped me out -- I was missing the .babelrc! All is well!

File: src/openmrs-esm-module-config.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    throw Error("No config schema has been defined for " + moduleName);
+  }
+  const schema = schemas[moduleName];
+
+  // Create a config object composed of all the defaults
+  const defaultConfig = R.map(R.prop('default'), schema);
+
+  // Merge all of the configs provided for moduleName
+  const allConfigsForModule = R.map(R.prop(moduleName), configs)
+  const providedConfig = R.mergeAll(allConfigsForModule);
```
Comment (joeldenning) [inline src/openmrs-esm-module-config.ts]: Looks like mergeAll enforces "last one wins"? Meaning the rightmost config in the array will override previous configs.

I'm assuming this is intended -- the last config you provide has more priority than the first?

File: src/openmrs-esm-module-config.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+  for (let [key, value] of Object.entries(providedConfig)) {
```
Comment (joeldenning) [inline src/openmrs-esm-module-config.ts]: This is not recursive and only checks top level properties. Are you planning on checking things deeper than top-level properties in a future PR? Or am I missing something about this code?

File: src/openmrs-esm-module-config.ts (hunk 7 above)
Comment (brandones) [reply src/openmrs-esm-module-config.ts]: Yep, exactly

File: src/openmrs-esm-module-config.ts (hunk 8 above)
Comment (brandones) [reply src/openmrs-esm-module-config.ts]: Right, uh, I haven't figured out how we should support nesting in the configs. I'll do something in the brainstorming doc. I'll let you know when I've added an idea. Feel free to pitch in there, if you have ideas about how it should happen.

Comment (brandones) [reply src/openmrs-esm-module-config.ts]: https://issues.openmrs.org/browse/MF-87

**Comment by brandones** [comment]
Thanks for the review(s), Joel!

**Pull request merged**